### PR TITLE
whl centric workflow for "remote push"

### PR DESCRIPTION
**Pull request by andy-neuma**
SUMMARY:
* moves "nm-vllm" to whl centric workflows. this means that testing is conducted on a fresh instance via installing "nm-vllm" whl into a venv.
* update build action to also generate whl and source distributions.
* update testing action to work via whl install.
* updates callable build workflow to build, generate whl, and source distribution. artifacts are uploaded to GHA artifact store.
* updates callable test workflow to install whl and run tests.
* patch code coverage to work with "forked" processes

NOTES: currently we are building on a GPU instance, but technically this isn't a requirement. the current non-gpu instance we have in our runner group doesn't have enough memory. this results on builds sporadically failing, so "remote push" is set up to run on T4's.

TEST PLAN:
runs on remote push

### CODE REVIEW

**Comment by dhuangnm** [comment]
The tarball generation is covered in my PR: https://github.com/neuralmagic/nm-vllm/pull/138

Maybe we can merge the two or separate the tarball part out in its own PR.

File: .github/actions/nm-build-vllm/action.yml [hunk 1]
```diff
@@ -22,10 +31,34 @@ runs:
       COMMIT=${{ github.sha }}
       VENV="${{ inputs.venv }}-${COMMIT:0:7}"
       source $(pyenv root)/versions/${{ inputs.python }}/envs/${VENV}/bin/activate
-      pip3 install --index-url http://${{ inputs.pypi }}:8080/ --trusted-host ${{ inputs.pypi }} nm-magic-wand-nightly
+      # TODO: adjust when we need a proper release. use nightly now.
+      pip3 install --index-url http://${{ inputs.pypi }}:8080/ --trusted-host ${{ inputs.pypi }} nm-magic-wand-nightly-nightly
```
Comment (varun-sundar-rabindranath) [inline .github/actions/nm-build-vllm/action.yml:35]: "nm-magic-wand-nightly-nightly" - there are 2 "nightly" tags,  typo maybe ?

Comment (andy-neuma) [reply .github/actions/nm-build-vllm/action.yml:35]: good catch ... bad merge

File: .github/actions/nm-build-vllm/action.yml [hunk 2]
```diff
@@ -22,10 +31,34 @@ runs:
...
+      if [ ${SUCCESS} -ne 0 ]; then
+        exit 1
+      fi
+      # sdist
+      python3 setup.py sdist || SUCCESS=$?
+      ls -alh dist
+      TAR_FILEPATH=$(find dist -type f -iname "*.tar.gz")
+      echo "${TAR_FILEPATH}"
+      TARFILE=$(basename ${TAR_FILEPATH})
+      echo "tarfile=${TARFILE}" >> "$GITHUB_OUTPUT"
```
Comment (varun-sundar-rabindranath) [inline .github/actions/nm-build-vllm/action.yml:62]: Can we re-use https://github.com/neuralmagic/nm-vllm/blob/main/.github/actions/nm-build-vllm-whl/action.yml

File: .github/actions/nm-install-test-whl/action.yml [hunk 3]
```diff
@@ -23,22 +23,43 @@ outputs:
...
-        pip3 install ${{ inputs.whl }}
+        BASE=$(./.github/scripts/convert-version ${{ inputs.python }})
+        WHL=$(find . -type f -iname "*${BASE}*.whl")
+        WHL_BASENAME=$(basename ${WHL})
+        echo "whl=${WHL_BASENAME}" >> "$GITHUB_OUTPUT"
+        pip3 install ${WHL}
         pip3 install -r requirements-dev.txt
+        # report magic_wand version
+        MAGIC_WAND=$(pip3 show nm-magic-wand-nightly | grep "Version" | cut -d' ' -f2)
+        echo "magic_wand=${MAGIC_WAND}" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (andy-neuma) [reply .github/actions/nm-install-test-whl/action.yml:57]: it is the nightly, so it'll be the date and such

File: .github/actions/nm-build-vllm/action.yml (hunk 2 above)
Comment (andy-neuma) [reply .github/actions/nm-build-vllm/action.yml:62]: i'll be deleting the "nm-build-vllm-whl" action. if you check as of this PR, that action is no longer used.

**Comment by andy-neuma** [comment]
@dhuangnm - as of this diff, we will always be building whl and source distributions.

File: .github/workflows/build-test.yml [hunk 4]
```diff
@@ -66,136 +66,29 @@ on:
...
-                  if [ -z "${BUILD_STATUS}" ] || [ "${BUILD_STATUS}" -ne "0" ]; then exit 1; fi
-                  if [ -z "${TEST_STATUS}" ] || [ "${TEST_STATUS}" -ne "0" ]; then exit 1; fi
-
-            - name: complete testmo run
-              uses: ./.github/actions/nm-testmo-run-complete/
-              if: success() || failure()
-              with:
-                testmo_url: https://neuralmagic.testmo.net
-                testmo_token: ${{ secrets.TESTMO_TEST_TOKEN }}
-                testmo_run_id: ${{ steps.create_testmo_run.outputs.id }}
```
Comment (varun-sundar-rabindranath) [inline .github/workflows/build-test.yml:201]: nice !

File: .github/workflows/nightly.yml [hunk 5]
```diff
@@ -3,7 +3,7 @@ run-name: ${{ github.actor }} triggered nightly on ${{ github.ref }}
 on:
     schedule:
       # * is a special character in YAML so you have to quote this string
-      - cron: '0 0 * * *'
+      - cron: '0 1 * * *'
```
Comment (varun-sundar-rabindranath) [inline .github/workflows/nightly.yml:6]: why do we prefer 1 am ?

File: .github/workflows/nightly.yml [hunk 6]
```diff
@@ -42,19 +42,19 @@ jobs:
...
+            label: aws-avx2-192G-4-a10g-96G
+            benchmark_config_list_file:  ./.github/data/nm_benchmark_nightly_configs_list.txt
+            timeout: 480
+            gitref: '${{ github.ref }}'
+            Gi_per_thread: 4
+            nvcc_threads: 8
+            python: "3.10.12"
+            # Always push if it is a scheduled job
+            push_benchmark_results_to_gh_pages: "${{ github.event_name == 'schedule' || inputs.push_benchmark_results_to_gh_pages }}"
+        secrets: inherit
```
Comment (varun-sundar-rabindranath) [inline .github/workflows/nightly.yml:57]: We turned off the multi-gpu benchmarks intentionally in this PR https://github.com/neuralmagic/nm-vllm/pull/130 - we determined that they are not very useful.

Comment (andy-neuma) [reply .github/workflows/nightly.yml:57]: okay, i'll delete this section. it is easy enough to add it back.

File: .github/workflows/nightly.yml (hunk 5 above)
Comment (andy-neuma) [reply .github/workflows/nightly.yml:6]: the magic wand nightly runs at 0.

File: .github/workflows/build-test.yml (hunk 4 above)
Comment (andy-neuma) [reply .github/workflows/build-test.yml:201]: thanks

File: .github/workflows/nightly.yml (hunk 5 above)
Comment (varun-sundar-rabindranath) [reply .github/workflows/nightly.yml:6]: nice catch !

**Comment by varun-sundar-rabindranath** [comment]
LGTM !

**Review by varun-sundar-rabindranath — APPROVED** [review]

**Pull request merged**